### PR TITLE
[SYCL-MLIR] Allow `unrealized_conversion_cast` operations in `-convert-polygeist-to-llvm`

### PR DESCRIPTION
**Pull request by victor-eds**
`unrealized_conversion_cast` operations are used in lowering to avoid inserting operations from other dialects. `-convert-polygeist-to-llvm` should be able to emit these and rely on a `-reconcile-unrealized-casts` pass to erase them.

### CODE REVIEW

**Review by jopperm — APPROVED** [review]
Nice clean-up!

**Comment by victor-eds** [comment]
`InvokeSimd/Feature/SPMD_invoke_ESIMD_external.cpp` failure unrelated to this.

**Review by sommerlukas — APPROVED** [review]

File: polygeist/test/polygeist-opt/sycl/sycl-constructors-to-llvm.mlir [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+// CHECK-NEXT:      %[[VAL_19:.*]] = llvm.mlir.constant(2 : i32) : i32
+// CHECK-NEXT:      %[[VAL_20:.*]] = llvm.getelementptr inbounds %[[VAL_14]][0, 0, 0, 2] : (!llvm.ptr) -> !llvm.ptr, !llvm.struct<"class.sycl::_V1::id.3", (struct<"class.sycl::_V1::detail::array.3", (array<3 x i64>)>)>
+// CHECK-NEXT:      llvm.store %[[VAL_2]], %[[VAL_20]] : i64, !llvm.ptr
+// CHECK-NEXT:      %[[VAL_21:.*]] = llvm.mlir.undef : !llvm.struct<(ptr, ptr, ptr)>
+// CHECK-NEXT:      %[[VAL_22:.*]] = llvm.insertvalue %[[VAL_4]], %[[VAL_21]][0] : !llvm.struct<(ptr, ptr, ptr)>
+// CHECK-NEXT:      %[[VAL_23:.*]] = llvm.insertvalue %[[VAL_8]], %[[VAL_22]][1] : !llvm.struct<(ptr, ptr, ptr)>
+// CHECK-NEXT:      %[[VAL_24:.*]] = llvm.insertvalue %[[VAL_14]], %[[VAL_23]][2] : !llvm.struct<(ptr, ptr, ptr)>
+// CHECK-NEXT:      llvm.return %[[VAL_24]] : !llvm.struct<(ptr, ptr, ptr)>
+// CHECK-NEXT:    }
+func.func @id_index(%arg0: index, %arg1: index, %arg2: index)
```
Comment (etiotto) [inline polygeist/test/polygeist-opt/sycl/sycl-constructors-to-llvm.mlir:79]: I am wondering whether these tests could be simplified and just check that the produced IR doe not contain any unrealized conversion casts ?

Comment (victor-eds) [reply polygeist/test/polygeist-opt/sycl/sycl-constructors-to-llvm.mlir:79]: I thought of these tests as a way to check `-convert-polygeist-to-llvm -reconcile-unrealized-casts` can handle operations whose lowering introduces such ops. This is more like a test for the whole lowering (should not be necessary if polygeist's lowering was well designed, but it relies on patterns from other conversion passes, so this integration testing is needed...)

**Review by whitneywhtsang — APPROVED** [review]

File: polygeist/test/polygeist-opt/sycl/sycl-constructors-to-llvm.mlir (hunk 1 above)
Comment (etiotto) [reply polygeist/test/polygeist-opt/sycl/sycl-constructors-to-llvm.mlir:79]: ok that is fine.

**Review by etiotto — APPROVED** [review]

**Pull request merged**